Add Core Functionality to Canvas Widget

danyeaw opened the pull request:
Context2D provides a 2D rendering context for drawing on a canvas and initial set of core drawing widgets. Matrix defines the transformation from user-space to device-space coordinates for advanced drawing transformations.

phildini commented on src/core/toga/widgets/canvas.py:
```diff
@@ -2,19 +2,21 @@
 
 class Canvas(Widget):
-    """"""
+    """Create new canvas
```
Am I missing something, or are the only changes to canvas docstring changes?

danyeaw replied:
Yes, that's correct, only the small change. I originally had the new Context2D and Matrix classes in the canvas module since they were all drawing related. I refactored them out after looking at some of the other modules. I thought having them in separate modules is more consistent with how the other widgets are implemented.

phildini replied:
Cool. I'm maybe -0 on changing this file just for docs changes, as it clutters the review and the blame history a bit, but I don't think it's a blocker necessarily.

I agree re: having the context be its own file.

freakboy3742 replied:
FWIW - I'm not so concerned about this. This PR is all about the Canvas widget; if implementing it has resulted in some re-organization or documentation updates, it's all "on topic" to my mind.

phildini commented:
Other than my comment above, I think this looks good.

Unless I missed it in the docs, how do you see `Context2D` and `Canvas` working together? Do we imagine there will someday be a `Context3D` that will also draw to the Canvas object?

danyeaw commented:
The different platforms do allow for different contexts, like a PDF or PS context. The name is based on the HTML canvas.

phildini commented:
Chatted with @danyeaw about this in gitter. I think this looks solid, the extensiveness of the testing helps, once this PR includes some example code snippets (examples folder in the docs, maybe?), I think it's fine to ship.

freakboy3742 commented on src/gtk/toga_gtk/widgets/context2d.py:
```diff
@@ -0,0 +1,105 @@
+import cairo
```
Does this call need to be wrapped in much the same way as the Webkit imports, to protect against platforms where (for whatever reason) Cairo isn't available?

danyeaw replied:
Yes I agree that this is a good idea. I committed a change to check for this using gi.require_foreign("cairo") which is the recommended method in the PyGObject user guide

freakboy3742 commented:
Oh - one more question: It's not clear to me that the Canvas API has an entry point that allows creation of the Context2D object. Is this an oversight, or have I missed something?

danyeaw commented:
@freakboy3742 Thanks for all of the great feedback! I'll work on reviewing and addressing your points.

danyeaw commented on src/core/toga/widgets/context2d.py:
```diff
@@ -0,0 +1,322 @@
+
+        Currently supports color, in the future could support gradient and
+        pattern. A named color or RGBA value must be passed, or default
+        to black.
```
I have made a commit that now supports the Colosseum representation (just needs to be uncommented once Colosseum is ready) and uses a reg exp to parse a string in the format 'rgba(1, 2, 3, 4)'

danyeaw commented on src/core/toga/widgets/context2d.py:
```diff
@@ -0,0 +1,322 @@
+
+    # Canvas State
+
+    def save(self):
```
Context manager added after doing some good study up, please let me know if this is what you had in mind.

danyeaw commented on src/core/toga/widgets/context2d.py:
```diff
@@ -0,0 +1,322 @@
+
+    # Paths
+
+    def begin_path(self):
```
Context manager added after doing some good study up, please let me know if this is what you had in mind.

danyeaw commented:
@freakboy3742 Thanks for all of the great feedback! I'm still thinking about whether keeping the Canvas and Matrix classes makes sense or if I can find another solution. My toolkit does derive a View based on using Gtk.DrawingArea as a base class and assigns specific Matrix's to an Item on that view. So I am trying to abstract some of these concepts for more wider use in Toga, but right now I'm struggling a bit. For a really simple use case in Toga, we could optionally just make a Canvas created with a Window, and then multiple contexts could be added as things are drawn. But, my toolkit keeps all of these views and items separate from a Window created in an app.

freakboy3742 commented:
The immediately obvious solution to me is that a canvas automatically gets a 2D context internally. There's no need to expose the context as a public API. 

This particular piece of API design hinges on one issue: What's the use case for having a single canvas, and more than one context (be it multiple 2D contexts, or a 2D and 3D context)? If there's a compelling reason to keep this, by by all means, we need to preserve it. If there isn't, it's an implementation complexity we can hide. If there is, but only in complex cases, then we can do something like have a default context, but allow other contexts to be created.

This might also be clearer if you can provide a sample app that uses this API. In the current structure, it still isn't clear to me how you get a 2D context that is tied to the canvas...

danyeaw commented on src/core/tests/widgets/test_context2d.py:
```diff
@@ -0,0 +1,126 @@
+        self.context.fill()
+
+    def test_release(self):
+        self.context.release()
```
It looks like using the Gtk.Widget.signal.draw that I can let Gtk+ worry about creating and releasing a context. I have deleted the release method.

danyeaw commented:
I have incorporated the feedback for another iteration. Context2D and Matrix are gone, and all of the drawing routines were added to Canvas. I added context managers around some of the common drawing routines. Work remaining: drawing on a Toga window isn't yet working. There are no errors, so I am trying to investigate why nothing is showing up. Once that is working, I'll attempt to update the tutorial/example to draw Tiberius, and add a lot of explanation :)
More feedback is of course appreciated!

freakboy3742 commented on docs/tutorial/index.rst:
```diff
@@ -45,4 +45,12 @@ a simple browser.
+Tutorial 4 - let's draw on a canvas! (Gtk+ only for now)
+===================================
+
+In :doc:`tutorial-4`, you will discover how to use the :mod:`toga.interface.widgets.context2D.Context2D` widget to draw
```
Similarly - Context2D->Canvas?

danyeaw replied:
Updated, thanks for catching that!

freakboy3742 commented on docs/tutorial/tutorial-4.rst:
```diff
@@ -0,0 +1,15 @@
+   :language: python
+
+
+In this example, we see a new Toga widget - :class:`.Canvas`.
```
We can wait until the API has settled down, but we probably need a bit more detail than just "see the code". A little background on how contexts & drawing work would be good. Doesn't have to be *incredibly* detailed, but a few pointers would be good.

danyeaw replied:
The tutorial now has an explanation on how the widget works.

freakboy3742 commented on src/core/tests/widgets/test_canvas.py:
```diff
@@ -12,3 +13,105 @@ def setUp(self):
+            self.testing_canvas.fill(preserve=True)
+            self.testing_canvas.stroke_style('rgba(0.25, 0.25, 0.25, 0.6)')
+            self.testing_canvas.line_width(1)
+            self.testing_canvas.stroke()
```
These tests validate that using the interface API doesn't raise errors, but it doesn't validate that the underlying API is actually invoked. The changes I added to the testing framework yesterday should allow you to validate the sequence of events that have occurred. You might need to add another assertion to the base TestCase; but validating, for example, that this test results in a context save, a series of draw operations, then a context restore would be helpful.

danyeaw replied:
I added a test_canvas_on_draw method to try to better cover these cases. I haven't yet added a complete sequence of events test with a new assertion yet.

freakboy3742 commented on src/core/toga/widgets/canvas.py:
```diff
@@ -1,20 +1,297 @@
+        self.rehint()
+
+    @contextmanager
+    def save_restore(self):
```
I'm not wild about the naming here. It definitely describes what will be done, but not why. `context()` might be a better name - it would allow for the following usage:
```
with canvas.context() as context:
     context.line(...)
```
Thoughts?

freakboy3742 commented on src/core/toga/widgets/canvas.py:
```diff
@@ -1,20 +1,297 @@
+        self._impl.restore()
+
+    def draw(self, draw_func):
+        self._impl.draw(draw_func)
```
If I draw without being inside a context manager, what is going to happen here? Would it make sense for every "draw directly on the canvas" operation to implicitly save and restore a context? Then, if you are *inside* a context manager, the creation of a new context can be skipped. Thoughts?

freakboy3742 commented on src/dummy/toga_dummy/factory.py:
```diff
@@ -35,6 +35,7 @@
     'Box',
     'Button',
     'Canvas',
+    'Context2D',
```
Is this still needed?

danyeaw replied:
I missed this one, updated, thanks!

freakboy3742 commented on src/core/toga/widgets/canvas.py:
```diff
@@ -1,20 +1,297 @@
+
+    @contextmanager
+    def begin_close_path(self, x, y):
+        """Creates a new path and then closes it
```
Another naming issue here; as a context manager, would it make more sense for this to be just `closed_path()`, so usage would be:
```
with canvas.closed_path(1,2):
    canvas.line(...)
    canvas.line(...)
```
?

danyeaw replied:
Makes sense, I renamed to closed_path. Thanks!

danyeaw commented on docs/reference/supported_platforms/Context2D.rst:
```diff
@@ -0,0 +1,16 @@
+    +-----------+----+-----+----+------+-----+-----+-------+
+    | Component |iOS |win32|web |django|cocoa| gtk |android|
+    +===========+====+=====+====+======+=====+=====+=======+
+    |Context2D| |no| ||no| ||no|||no|  ||no| ||yes|||no|   |
```
Thanks, removed Context2D.rst

danyeaw commented:
Iteration 3 has incorporated most of the feedback: I still need to keep working on test cases and figure out why drawing on the toga window isn't actually showing anything. I would appreciate another review, and then I'll try to solve hopefully the last piece of this puzzle.

danyeaw commented on examples/button/button/app.py:
```diff
@@ -70,7 +70,7 @@ def callbackLabel(self, button):
         # Some action when you hit the button
         #   In this case the window size will change
-        self.main_window.size = (100,100)
+        self.main_window.size(100,100)
```
Whoops, undid this accidental change.

danyeaw commented on docs/tutorial/tutorial-4.rst:
```diff
@@ -0,0 +1,41 @@
+
+2. Next we create a new Canvas, and pass in the draw_steps method::
+
+    self.canvas = toga.Canvas(on_draw=self.draw_steps)
```
It now should be possible to draw directly without wrapping the operations in a function.

danyeaw commented:
Tiberius says it's ready to ship!

![tutorial-4](https://user-images.githubusercontent.com/10014976/33226987-c0a6cd3a-d167-11e7-9d56-2e11c2b63df9.png)
